enhancements/kube-apiserver/startup-monitor.md

```diff
@@ -0,0 +1,234 @@
+
+The fallback mechanism is opt-in by the operator, and kube-apiserver and etcd operators (and potentially kcm and ks) have to consult the deployment topology in the infrastructure resource to decide. Most of the text here is generic, but we will focus mainly on the cluster-kube-apiserver-operator use-case first.
+
+## Motivation
+
+In SNO, we only have one chance to start the new static pod for the kube-apiserver or etcd. When this fails, the cluster is bricked and there is no automatic recovery.
+
+Bad configurations can have many reasons:
+- **the pod manifest can be wrong**: e.g. by YAML syntax, not validating as a pod or some deeper semantical error:
+  - **each one of the [23 different config observers of kube-apiserver]([https://](https://github.com/openshift/cluster-kube-apiserver-operator/blob/ce4170b4a040fc03603f62d686a6e9ea0cadde34/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L119)) provided and owned by 9 different teams can lead to bad configuration**.
+    E.g. a config observer adds a flag `--foo` which was removed in the current upstream release, but lacking test coverage did not show this in CI.
+  - **source invalid data from other sources with incomplete validation**: e.g. the network mask given through some other config.openshift.io/v1 CR defined by the user is invalid.
+- **some ConfigMap or Secret can be wrong**: e.g. the audit policy is a YAML file inside a ConfigMap. If it is invalid, the kube-apiserver will refuse startup. ConfigMaps and Secrets are synced into the operand namespace from many different sources, and then copied as files onto the master node file system by the installer pod.
+- **some ConfigMap or Secret marked as optional but is actually required**: the installer will continue rolling out the pod after skipping the optional files. In HA OpenShift, this might not show up because the operator create quickly another revision when the ConfigMap or Secret show up. In SNO this race might be fatal.
+
+It is impossible to guarantee that none of these ever happens because the test permutations are just infeasible to cover completely in CI, especially as many of these are disruptive for cluster behaviour and hence very expensive to test.
+
+### Goals
+
+- recover from a badly formatted or invalid pod manifest
+- recover from bad config observer output
+- recover from bad revisioned ConfigMaps and Secrets.
+
+### Non-Goals
+
+- recover from invalid non-revisioned certificates
+- recover from missing or invalid files on the masters given by static paths outside of the `/etc/kubernetes` directory (e.g. some certs or kubeconfig the operand is consuming)
+- optimize downtime of the API beyond in fallback case to O(60s). This mechanism is for the disaster case and only survival counts. If downtime is `O(5min)` or even more, this is completely ok.
```

Comment:
Pod manifest validation seems like a reasonable addition to the operator, although I wonder if this should be covered here as well.

Reply:
dry-run can catch a reasonable amount of these I would say. Though I bet it's not complete.